thrust/system/detail/generic/shuffle.inl

```diff
@@ -32,94 +32,90 @@ namespace system {
-      key[i] = g();
-    }
+  __host__ __device__ rc5_bijection(uint64_t m, URBG &&g)
+   : w(get_cipher_bits(m))
+   , w_is_pow_2((w & (w-1)) == 0)
+  {
+    init_state(std::forward<URBG>(g));
   }
 
-  __host__ __device__ uint64_t nearest_power_of_two() const {
-    return 1ull << (left_side_bits + right_side_bits);
+  __host__ __device__ uint64_t bijection_width() const {
+    return 1ull << (2*w);
   }
-  __host__ __device__ uint64_t operator()(const uint64_t val) const {
-    // Extract the right and left sides of the input
-    uint32_t left = (uint32_t)(val >> right_side_bits);
-    uint32_t right = (uint32_t)(val & right_side_mask);
-    round_state state = {left, right};
 
-    for (uint64_t i = 0; i < num_rounds; i++) {
+  __host__ __device__ uint64_t operator()(const uint64_t val) const {
+    assert(val < bijection_width());
+    if(w == 0)
+      return val;
+    round_state state = { (uint32_t)val & get_mask(), (uint32_t)(val >> w) };
+    state.A = (state.A + S[0]) & get_mask();
+    state.B = (state.B + S[1]) & get_mask();
+    for(uint32_t i = 0; i < num_rounds; i++)
       state = do_round(state, i);
-    }
-
-    // Check we have the correct number of bits on each side
-    assert((state.left >> left_side_bits) == 0);
-    assert((state.right >> right_side_bits) == 0);
+    uint64_t res = state.B << w | state.A;
+    assert(res < bijection_width());
+    return res;
+  }
 
-    // Combine the left and right sides together to get result
-    return state.left << right_side_bits | state.right;
+private:
+  template <class URBG>
+  __host__ __device__ void init_state(URBG&& g)
+  {
+    thrust::uniform_int_distribution<uint32_t> dist(0, get_mask());
+    for( uint32_t i = 0; i < state_size; i++ )
+      S[i] = dist(g);
   }
 
- private:
   // Find the nearest power of two
   __host__ __device__ uint64_t get_cipher_bits(uint64_t m) {
+    if(m == 0)
+      return 0;
     uint64_t i = 0;
+    m--;
     while (m != 0) {
       i++;
-      m >>= 1;
+      m >>= 2u;
     }
     return i;
   }
 
-  // Round function, a 'pseudorandom function' whos output is indistinguishable
-  // from random for each key value input. This is not cryptographically secure
-  // but sufficient for generating permutations. We hash the value with the
-  // tau88 engine and combine it with the random bits of the key (provided by
-  // the user-defined engine).
-  __host__ __device__ uint32_t round_function(uint64_t value,
-                                              const uint64_t key) const {
-    uint64_t value_hash = thrust::random::taus88(value)();
-    return (value_hash ^ key) & left_side_mask;
+  __host__ __device__ uint32_t get_mask() const
+  {
+    return (uint32_t)((1ull << (uint64_t)w) - 1ull);
+  }
+
+  __host__ __device__ uint32_t rotl( uint32_t val, uint32_t amount ) const
+  {
+    assert((val & ~get_mask()) == 0);
+    const uint32_t amount_mod = w_is_pow_2 ? (amount & (w-1)) : (amount % w);
```

Comment:
Is this a performance optimisation? If so, I would just use %. GPU code does not like branches. If the compiler uses predication, both branches get executed always.

Reply:
GPU code doesnt like branches if the warp/half warp diverges. In this case it will always take one branch. The branch predictor should also be able to predict the correct branch since again - always takes the same.
My understanding is that while GPUs dont like branches they like modulus/division even less.

Reply:
It's likely that ternary operators uses predication instead of prediction. We are just speculating without looking at sass code in a profiler, but in my experience this is not an effective optimisation.

Reply:
Ah so predication has the latency of both instructions even if no thread in the warp takes the branch?